advanced_server.rb

```diff
@@ -0,0 +1,314 @@
+          file_path = file["path"].gsub(/#{repository}\//,'')
+          blob_href = "#{repository_url}/blob/#{head_sha}/#{file_path}"
+          warning_level = 'notice'
+          file["offenses"].each do |offense|
+            start_line   = offense["location"]["start_line"]
+            end_line     = offense["location"]["last_line"]
+            start_column = offense["location"]["start_column"]
+            end_column   = offense["location"]["last_column"]
+            message      = offense["message"]
+            annotation = {
+              "filename" => file_path,
+              "blob_href" => blob_href,
+              "start_line" => start_line,
+              "end_line" => end_line,
+              "start_column" => start_column,
+              "end_column" => end_column,
+              "warning_level" => warning_level,
+              "message" => message
+            }
+            annotations.push(annotation)
+          end
+        end
+      end
+
+      summary = "Octoc Rubocop summary\n-Offense count: #{@output["summary"]["offense_count"]}\n-File count: #{@output["summary"]["target_file_count"]}\n-Target file count: #{@output["summary"]["inspected_file_count"]}}"
+      details = "Octoc Rubocop version: #{@output["metadata"]["rubocop_version"]}"
+
+      # Now, mark the check run as complete! And if there are warnings, share them
+      updated_check_run = @installation_client.patch("repos/#{@payload['repository']['full_name']}/check-runs/#{@payload['check_run']['id']}", {
+          accept: 'application/vnd.github.antiope-preview+json', # This header is necessary for beta access to Checks API
+          name: 'Octo Rubocop',
+          status: 'completed',
+          conclusion: conclusion,
+          completed_at: Time.now.utc.iso8601,
+          output: {
+            title: "Octoc Rubocop",
+            summary: summary,
+            details: details,
+            annotations: annotations
+          },
+          actions: [{
+            label: "Fix this",
+            description: "Automatically fix all linter notices.",
+            identifier: "fix_rubocop_notices"
+          }]
+      })
+
+    end
+
+    def take_requested_action
+      full_repo_name = @payload['repository']['full_name']
+      repository     = @payload['repository']['name']
+      head_sha       = @payload['check_run']['head_sha']
+      head_branch    = @payload['check_run']['check_suite']['head_branch']
+      repository_url = @payload['repository']['html_url']
+
+      if(@payload['requested_action']['identifier'] == 'fix_rubocop_notices')
+        clone_repository(full_repo_name, repository, head_sha, head_branch)
+
+        @report = `rubocop '#{repository}/*' --format json --auto-correct`
+        pwd = Dir.getwd()
+        Dir.chdir(repository)
+        `git add .`
+        `git commit -am 'Automatically fix Octo Rubocop notices.'`
+        `git push 'https://github.com/#{full_repo_name}.git' #{head_branch}`
```

Comment:
@DEGoodmanWilson I really wanted to find a linter that I could avoid command line calls with, but I didn't have any luck. 

> I feel like there is a sequence of API calls that we should be using here to ensure this actually succeeds.

Are you referring to the https://developer.github.com/v3/git/ API? Originally, I planned to use this API but didn't think it would work when running the linter on the command line.

Reply:
You'll need to authenticate as your installation for this to work: https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#http-based-git-access-by-an-installation

Reply:
The app will already have authenticated as an installation, that's no problem. The question I think @swinton is how can we `git push` code from disk using that installation key?

Reply:
@swinton @DEGoodmanWilson I've already authenticated when I clone the repo using [this](https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#http-based-git-access-by-an-installation) method. Do I still need to authenticate when I push code?

Reply:
I'm not :100:, I _think_ you'll need to provide the token again when you push, i.e. 

```ruby
`git push 'https://x-access-token:#{@installation_token.to_s}@github.com/#{full_repo_name}.git' #{head_branch}`
```

Reply:
I don't believe the access token works on push. I reached out in slack about how to push code from the command line using a GitHub App. I ended adding some code to update the config with a user and email, so that user is the committer/author shown on GitHub. Here is the reference slack [convo](https://github.slack.com/archives/CD9PT6ZDY/p1543264767664500).

Reply:
This seems a huge improvement 💃